lib/puppet/type/service.rb

```diff
@@ -230,5 +230,31 @@ def refresh
         debug "Skipping restart; service is not running"
       end
     end
+
+    def retrieve
+      fail "Provider #{provider.class.name} is not functional on this host" if self.provider.is_a?(Puppet::Provider) and ! provider.class.suitable?
+
+      result = Puppet::Resource.new(self.class, title)
+
+      # Provide the name, so we know we'll always refer to a real thing
+      result[:name] = self[:name] unless self[:name] == title
+
+      if ensure_prop = property(:ensure)
+        result[:ensure] = ensure_state = ensure_prop.retrieve
+      else #Don't check anything if there is no ensure property
+        ensure_state = nil
+      end
+
+      properties.each do |property|
+        next if property.name == :ensure
+        if ensure_state == :absent
+          result[property] = :absent
+        else
+          result[property] = property.retrieve
+        end
+      end
+
+      result
+    end
```

Comment:
Note to reviewers, the is mostly copy and pasted from https://github.com/puppetlabs/puppet/blob/3.x/lib/puppet/type.rb#L1039-L1063. The difference is the superclass `Puppet::Type#retrieve` does:

```
if ensure_prop = property(:ensure) or (self.class.validattr?(:ensure) and ensure_prop = newattr(:ensure))
```

But this PR changes the service type only to do:

```
 if ensure_prop = property(:ensure)
```

The PR modifies puppet to behave like I thought it did, only manage the state of resources that are expressed in the manifest. For some reason, the `ensure` property is handled special, and the behavior was added in 3f6c9481.

When is it necessary to evaluate the second half? Isn't it always true for `ensurable` types? I'm wondering if we can make this change in `Puppet::Type` instead of special casing the service type.